lib/jxl/base/data_parallel.h

```diff
@@ -104,21 +97,34 @@ class ThreadPool {
   void* const runner_opaque_;
 };
 
-// TODO(deymo): Convert the return value to a Status when not using SkipInit.
 template <class InitFunc, class DataFunc>
-bool RunOnPool(ThreadPool* pool, const uint32_t begin, const uint32_t end,
-               const InitFunc& init_func, const DataFunc& data_func,
-               const char* caller) {
-  Status ret = true;
+Status RunOnPool(ThreadPool* pool, const uint32_t begin, const uint32_t end,
+                 const InitFunc& init_func, const DataFunc& data_func,
+                 const char* caller) {
+  if (pool == nullptr) {
+    ThreadPool default_pool(nullptr, nullptr);
+    return default_pool.Run(begin, end, init_func, data_func, caller);
+  } else {
+    return pool->Run(begin, end, init_func, data_func, caller);
+  }
+}
+template <class DataFunc>
+void RunOnPoolSkipInit(ThreadPool* pool, const uint32_t begin,
+                       const uint32_t end, const DataFunc& data_func,
+                       const char* caller) {
   if (pool == nullptr) {
     ThreadPool default_pool(nullptr, nullptr);
-    ret = default_pool.Run(begin, end, init_func, data_func, caller);
+    JXL_CHECK(default_pool.Run(begin, end, ThreadPool::TrivialInit, data_func,
```

Comment:
So... how do we know that the no-init version cannot fail? Does the C API for threading state that this is the case?

Reply:
Only the init returns something. The data function has a void return type.

I don't know what should happen if someone implements an incorrect JxlParallelRunner that returns failure even if the init function returned success. This check will then fail.

Reply:
What if the parallel runner fails for internal reasons?

Reply:
It's not clear to me what the actual api is for parallel runners. If it can fail for internal reasons, then we have a bunch of places where this is not handled (none of the calls to RunOnPoolSkipInit check the result). So I was assuming that this cannot happen...

Reply:
Then the correct solution would be to handle those errors, or to clarify in the documentation of the thread pool API that this should not happen.

Reply:
Which option do you prefer? Currently it looks like the thread pool API allows a runner to fail for other reasons than the init failing, but none of our code with SkipInit is handling that possibility.

I'm inclined to say that the thread pool API should be changed so it cannot fail if init succeeds - otherwise it would be a technically valid implementation to just always fail, or to have unreasonable requirements like only working if 256 threads are available. I think it's better to make it so that a thread runner can only return failure if init failed, even if that means it has to fall back to just running things sequentially.